remill/Arch/X86/Semantics/MMX.h

```diff
@@ -73,6 +149,784 @@ DEF_SEM(PUNPCKHQDQ, D dst, S1 src1, S2 src2) {
+DEF_SEM(PADDB, D dst, S1 src1, S2 src2) {
+	auto lhs_vec = UReadV8(src1);
+	auto rhs_vec = UReadV8(src2);
+	auto dst_vec = UAddV8(lhs_vec, rhs_vec);
+	UWriteV8(dst, dst_vec);
+}
+
+DEF_ISEL(PADDB_MMXq_MMXq) = PADDB<V64W, V64, V64>;
+DEF_ISEL(PADDB_MMXq_MEMq) = PADDB<V64W, V64, MV64>;
+
+
+template <typename D, typename S1, typename S2>
+DEF_SEM(PADDW, D dst, S1 src1, S2 src2) {
+	auto lhs_vec = UReadV16(src1);
+	auto rhs_vec = UReadV16(src2);
+	auto dst_vec = UAddV16(lhs_vec, rhs_vec);
+	UWriteV16(dst, dst_vec);
+}
+
+DEF_ISEL(PADDW_MMXq_MMXq) = PADDW<V64W, V64, V64>;
+DEF_ISEL(PADDW_MMXq_MEMq) = PADDW<V64W, V64, MV64>;
+
+template <typename D, typename S1, typename S2>
+DEF_SEM(PADDD, D dst, S1 src1, S2 src2) {
+	auto lhs_vec = UReadV32(src1);
+	auto rhs_vec = UReadV32(src2);
+	auto dst_vec = UAddV32(lhs_vec, rhs_vec);
+	UWriteV32(dst, dst_vec);
+}
+
+DEF_ISEL(PADDD_MMXq_MMXq) = PADDD<V64W, V64, V64>;
+DEF_ISEL(PADDD_MMXq_MEMq) = PADDD<V64W, V64, MV64>;
+
+template <typename D, typename S1, typename S2>
+DEF_SEM(PADDQ, D dst, S1 src1, S2 src2) {
+	auto lhs_vec = UReadV64(src1);
+	auto rhs_vec = UReadV64(src2);
+	auto dst_vec = UAddV64(lhs_vec, rhs_vec);
+	UWriteV64(dst, dst_vec);
+}
+
+DEF_ISEL(PADDQ_MMXq_MMXq) = PADDQ<V64W, V64, V64>;
+DEF_ISEL(PADDQ_MMXq_MEMq) = PADDQ<V64W, V64, MV64>;
+
+/*
+ * signed saturation arithematic for PADDS
+ * int8 res = a + b;
+ * int8 tmp = (res & ~(a | b)) < 0 ? 0x7f : res;
+ * int8 c = (~res & (a & b)) < 0 ? 0x80 : tmp;
+ *
+ */
+
+template <typename D, typename S1, typename S2>
+DEF_SEM(PADDSB, D dst, S1 src1, S2 src2) {
+	auto src1_vec = SReadV8(src1);
+	auto src2_vec = SReadV8(src2);
+	auto add_vec = SAddV8(src1_vec, src2_vec);
+	auto or_vec = SOrV8(src1_vec, src2_vec);
+	auto and_vec = SAndV8(src1_vec, src2_vec);
+
+	//Compute signed saturation arithematic on each bytes
+    auto dst_vec = SClearV8(SReadV8(dst));
+	for (size_t index = 0; index < NumVectorElems(add_vec); index++) {
+		auto add_elem = SExtractV8(add_vec, index);
+		auto or_elem = SExtractV8(or_vec, index);
+		auto and_elem = SExtractV8(and_vec, index);
+		auto tmp = Select(SCmpLt(SAnd(add_elem, SNot(or_elem)), decltype(add_elem)(0)), decltype(add_elem)(0x7F), add_elem);
+		auto value = Select(SCmpLt(SAnd(SNot(add_elem), and_elem), decltype(add_elem)(0)), decltype(add_elem)(0x80), tmp);
+		dst_vec = SInsertV8(dst_vec, index, value);
+	}
+	SWriteV8(dst, dst_vec);
+}
+
+DEF_ISEL(PADDSB_MMXq_MMXq) = PADDSB<V64W, V64, V64>;
+DEF_ISEL(PADDSB_MMXq_MEMq) = PADDSB<V64W, V64, MV64>;
+
+
+template <typename D, typename S1, typename S2>
+DEF_SEM(PADDSW, D dst, S1 src1, S2 src2) {
+	auto src1_vec = SReadV16(src1);
+	auto src2_vec = SReadV16(src2);
+	auto add_vec = SAddV16(src1_vec, src2_vec);
+	auto or_vec = SOrV16(src1_vec, src2_vec);
+	auto and_vec = SAndV16(src1_vec, src2_vec);
+
+	//Compute signed saturation arithematic on each bytes
+    auto dst_vec = SClearV16(SReadV16(dst));
+	for (size_t index = 0; index < NumVectorElems(add_vec); index++) {
```

Comment:
I think it will not be able to handle byte saturation in-case of overflow (e.g, v1 = 0x40, v2= 0x7E); I will improve on handling the loop;